### ISSUE TITLE
[BUG] Java Mission Control shows many NoSuchMethodError thrown while building blob client.

### ISSUE THREAD

**Issue by krystiangorecki** [Azure/azure-sdk-for-java#28848]
**Describe the bug**
Java Flight Recorder reports tremendous amount of NoSuchMethodError thrown while running simple hello world code to access blob storage. However application logic works as expected. **My only dependency is com.azure:azure-storage-blob:12.16.0**  
The message is:
```
The program generated an average of 66 errors per minute during 5/12/2022, 3:18:59.000 PM – 3:19:59 PM.
66 errors were thrown in total. The most common error was ''java.lang.NoSuchMethodError'', which was thrown 56 times. 
```

***Exception or Stack Trace***
One of many similar occurrences copied from JDK Mission Control:
```
Stack Trace
void java.lang.Error.<init>(String)
void java.lang.LinkageError.<init>(String)
void java.lang.IncompatibleClassChangeError.<init>(String)
void java.lang.NoSuchMethodError.<init>(String)
java.lang.invoke.MemberName java.lang.invoke.MethodHandleNatives.resolve(MemberName, Class, boolean)
java.lang.invoke.MemberName java.lang.invoke.MemberName$Factory.resolve(byte, MemberName, Class, boolean)
java.lang.invoke.MemberName java.lang.invoke.MemberName$Factory.resolveOrNull(byte, MemberName, Class)
java.lang.invoke.MemberName java.lang.invoke.InvokerBytecodeGenerator.resolveFrom(String, MethodType, Class)
java.lang.invoke.MemberName java.lang.invoke.InvokerBytecodeGenerator.lookupPregenerated(LambdaForm, MethodType)
java.lang.invoke.MemberName java.lang.invoke.InvokerBytecodeGenerator.generateCustomizedCode(LambdaForm, MethodType)
void java.lang.invoke.LambdaForm.compileToBytecode()
java.lang.invoke.LambdaForm java.lang.invoke.DirectMethodHandle.makePreparedLambdaForm(MethodType, int)
java.lang.invoke.LambdaForm java.lang.invoke.DirectMethodHandle.preparedLambdaForm(MethodType, int)
java.lang.invoke.LambdaForm java.lang.invoke.DirectMethodHandle.preparedLambdaForm(MemberName, boolean)
java.lang.invoke.LambdaForm java.lang.invoke.DirectMethodHandle.preparedLambdaForm(MemberName)
java.lang.invoke.DirectMethodHandle java.lang.invoke.DirectMethodHandle.make(byte, Class, MemberName, Class)
java.lang.invoke.MethodHandle java.lang.invoke.MethodHandles$Lookup.getDirectMethodCommon(byte, Class, MemberName, boolean, boolean, MethodHandles$Lookup)
java.lang.invoke.MethodHandle java.lang.invoke.MethodHandles$Lookup.getDirectMethodNoSecurityManager(byte, Class, MemberName, MethodHandles$Lookup)
java.lang.invoke.MethodHandle java.lang.invoke.MethodHandles$Lookup.getDirectMethodForConstant(byte, Class, MemberName)
java.lang.invoke.MethodHandle java.lang.invoke.MethodHandles$Lookup.linkMethodHandleConstant(byte, Class, String, Object)
java.lang.invoke.MethodHandle java.lang.invoke.MethodHandleNatives.linkMethodHandleConstant(Class, int, Class, String, Object)
void reactor.netty.ReactorNetty.<clinit>()
reactor.netty.ChannelPipelineConfigurer reactor.netty.ChannelPipelineConfigurer.emptyConfigurer()
void reactor.netty.transport.TransportConfig.<init>(Map)
void reactor.netty.transport.ClientTransportConfig.<init>(ConnectionProvider, Map, Supplier)
void reactor.netty.http.client.HttpClientConfig.<init>(ConnectionProvider, Map, Supplier)
void reactor.netty.http.client.HttpClientConnect.<init>(ConnectionProvider)
reactor.netty.http.client.HttpClient reactor.netty.http.client.HttpClient.create()
com.azure.core.http.HttpClient com.azure.core.http.netty.NettyAsyncHttpClientBuilder.build()
com.azure.core.http.HttpClient com.azure.core.http.netty.NettyAsyncHttpClientProvider.createInstance()
com.azure.core.http.HttpClient com.azure.core.implementation.http.HttpClientProviders.createInstance(ClientOptions)
com.azure.core.http.HttpClient com.azure.core.http.HttpClient.createDefault(HttpClientOptions)
com.azure.core.http.HttpClient com.azure.core.http.HttpClient.createDefault()
com.azure.core.http.HttpPipeline com.azure.core.http.HttpPipelineBuilder.build()
com.azure.core.http.HttpPipeline com.azure.storage.blob.implementation.util.BuilderHelper.buildPipeline(StorageSharedKeyCredential, TokenCredential, AzureSasCredential, String, String, RequestRetryOptions, HttpLogOptions, ClientOptions, HttpClient, List, List, Configuration, ClientLogger)
com.azure.storage.blob.BlobServiceAsyncClient com.azure.storage.blob.BlobServiceClientBuilder.buildAsyncClient()
com.azure.storage.blob.BlobServiceClient com.azure.storage.blob.BlobServiceClientBuilder.buildClient()
void com.example.App.main(String[])
```

**To Reproduce**
Run included snippet with JVM option `-XX:StartFlightRecording:filename=/home/xxx/record.jfr`
Open recorded file with [JDK Mission Control](https://www.oracle.com/java/technologies/javase/products-jmc8-downloads.html) to see detailed analysis.

***Code Snippet***
```
    <dependencies>
        <dependency>
            <groupId>com.azure</groupId>
            <artifactId>azure-storage-blob</artifactId>
            <version>12.16.0</version>
        </dependency>
    </dependencies>
```

```
    public static void main(String[] args) {

        BlobServiceClient blobServiceClient = new BlobServiceClientBuilder()
            .credential(new StorageSharedKeyCredential(ACCOUNT_NAME, ACCESS_KEY))
            .endpoint(ENDPOINT_URL)
            .buildClient();

        BlobContainerClient blobContainerClient = blobServiceClient.getBlobContainerClient(CONTAINER);

        BlobClient blobClient = blobContainerClient.getBlobClient("aaa.txt");
        System.out.println(blobClient.exists());
    }
```

**Screenshots**
![Screenshot from 2022-05-12 15-31-00](https://user-images.githubusercontent.com/10375973/168245149-0cb64147-a91d-4381-8a17-93cb6d89c23a.png)

**Setup**
 - OS: Ubuntu 20.04.4 LTS
 - IDE: IntelliJ Idea
 - Library/Libraries: com.azure:azure-storage-blob:12.16.0  
 - Java version: openjdk version "11.0.15" 2022-04-19, OpenJDK Runtime Environment (build 11.0.15+10-Ubuntu-0ubuntu0.20.04.1)

**mvn dependency:tree**
```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ azure-poc-blob-minimal ---
[INFO] com.example:azure-poc-blob-minimal:jar:1.0-SNAPSHOT
[INFO] \- com.azure:azure-storage-blob:jar:12.16.0:compile
[INFO]    +- com.azure:azure-core:jar:1.27.0:compile
[INFO]    |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.13.2:compile
[INFO]    |  +- com.fasterxml.jackson.core:jackson-core:jar:2.13.2:compile
[INFO]    |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.13.2.1:compile
[INFO]    |  +- com.fasterxml.jackson.datatype:jackson-datatype-jsr310:jar:2.13.2:compile
[INFO]    |  +- com.fasterxml.jackson.dataformat:jackson-dataformat-xml:jar:2.13.2:compile
[INFO]    |  |  +- org.codehaus.woodstox:stax2-api:jar:4.2.1:compile
[INFO]    |  |  \- com.fasterxml.woodstox:woodstox-core:jar:6.2.7:compile
[INFO]    |  +- org.slf4j:slf4j-api:jar:1.7.33:compile
[INFO]    |  +- io.projectreactor:reactor-core:jar:3.4.14:compile
[INFO]    |  |  \- org.reactivestreams:reactive-streams:jar:1.0.3:compile
[INFO]    |  \- io.netty:netty-tcnative-boringssl-static:jar:2.0.47.Final:compile
[INFO]    |     \- io.netty:netty-tcnative-classes:jar:2.0.47.Final:compile
[INFO]    +- com.azure:azure-core-http-netty:jar:1.11.9:compile
[INFO]    |  +- io.netty:netty-handler:jar:4.1.73.Final:compile
[INFO]    |  |  +- io.netty:netty-common:jar:4.1.73.Final:compile
[INFO]    |  |  +- io.netty:netty-resolver:jar:4.1.73.Final:compile
[INFO]    |  |  +- io.netty:netty-transport:jar:4.1.73.Final:compile
[INFO]    |  |  \- io.netty:netty-codec:jar:4.1.73.Final:compile
[INFO]    |  +- io.netty:netty-handler-proxy:jar:4.1.73.Final:compile
[INFO]    |  |  \- io.netty:netty-codec-socks:jar:4.1.73.Final:compile
[INFO]    |  +- io.netty:netty-buffer:jar:4.1.73.Final:compile
[INFO]    |  +- io.netty:netty-codec-http:jar:4.1.73.Final:compile
[INFO]    |  +- io.netty:netty-codec-http2:jar:4.1.73.Final:compile
[INFO]    |  +- io.netty:netty-transport-native-unix-common:jar:4.1.73.Final:compile
[INFO]    |  +- io.netty:netty-transport-native-epoll:jar:linux-x86_64:4.1.73.Final:compile
[INFO]    |  |  \- io.netty:netty-transport-classes-epoll:jar:4.1.73.Final:compile
[INFO]    |  +- io.netty:netty-transport-native-kqueue:jar:osx-x86_64:4.1.73.Final:compile
[INFO]    |  |  \- io.netty:netty-transport-classes-kqueue:jar:4.1.73.Final:compile
[INFO]    |  \- io.projectreactor.netty:reactor-netty-http:jar:1.0.15:compile
[INFO]    |     +- io.netty:netty-resolver-dns:jar:4.1.72.Final:compile
[INFO]    |     |  \- io.netty:netty-codec-dns:jar:4.1.72.Final:compile
[INFO]    |     +- io.netty:netty-resolver-dns-native-macos:jar:osx-x86_64:4.1.72.Final:compile
[INFO]    |     |  \- io.netty:netty-resolver-dns-classes-macos:jar:4.1.72.Final:compile
[INFO]    |     \- io.projectreactor.netty:reactor-netty-core:jar:1.0.15:compile
[INFO]    +- com.azure:azure-storage-common:jar:12.15.1:compile
[INFO]    \- com.azure:azure-storage-internal-avro:jar:12.2.1:compile
```
**Information Checklist**
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

**Comment by alzimmermsft** [comment]
Thanks for filing this @krystiangorecki, after debugging through what is causing the `NoSuchMethodError`s to be thrown, it appears most are stemming from the JVM attempting to create `MethodHandle`s for lambda references such as `String::valueOf` or lambda functions.

**Comment by alzimmermsft** [comment]
Apologies on the delay @krystiangorecki.

I don't think there is anything being done wrong as long as your application functions as you expect. For example, I wrote this short snippet:

```java
public static void main(String[] args) {
    for (int i = 0; i < 10000; i++) {
        try {
            throw new NoSuchMethodError("Example");
        } catch (NoSuchMethodError error) {
            System.out.println("Caught a NoSuchMethodError that doesn't affect runtime, continuing.");
        }
    }
}
```

Where when I run it with flight recording enabled. After it finishes, I inspect the recording it says there were 10,000 `NoSuchMethodError`s thrown, but I know nothing is wrong as I expected that behavior. A real example of this can be found in the Azure SDKs, https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core/src/main/java/com/azure/core/implementation/jackson/XmlMapperFactory.java#L48, where we reflectively check for a Jackson API only available in Jackson 2.12+ but we support Jackson 2.11 and have a fallback for that. If Jackson 2.11 is used this will throw a `NoSuchMethodError`, or some form of `LinkageError`, but we planned for that and will continue running as expected.
